### ISSUE TITLE
[BUG] Cython performance regression with untyped NumPy array

### ISSUE THREAD

**Issue by rhshadrach** [cython/cython#5719]
### Describe the bug

We're attempting to upgrade to Cython 3.0.x in https://github.com/pandas-dev/pandas/pull/55179, but seeing some performance regressions. We've tracked a number of these down to our use of `ndarray` without any dtype or dimension specified. Is this change in performance expected?

Where we can, we've added the dtype and seen good performance improvements, but there are number of places in our Cython code that need to be able to handle an arbitrar NumPy array.

### Code to reproduce the behaviour:

main.pyx:

```cython
cimport numpy as cnp
from numpy cimport ndarray

cnp.import_array()

def foo(ndarray arr):
    cdef int n, i
    n = len(arr)
    for i in range(n):
        arr[i] += 1
```

run.py:

```
import numpy as np
from main import foo
import time

arr = np.arange(10_000_000)
timer = time.time()
foo(arr)
print(time.time() - timer)
```

setup.py:

```
from setuptools import setup
from Cython.Build import cythonize
import numpy as np

setup(
    ext_modules=cythonize("main.pyx"),
    include_dirs=[np.get_include()]
)
```

Build command:

```
python setup.py build_ext --inplace
```

### Expected behaviour

When compiling with Cython 0.29.33, the above code takes ~0.63 seconds to run. On Cython 3.0.2, it's ~1.04.

### OS

Linux

### Python version

3.10.12

### Cython version

3.0.2 vs 0.29.33

**Comment by matusvalo** [comment]
I can confirm the numbers. Checked also against master.

**Comment by matusvalo** [comment]
I had a look into the example and I have some comments. But note that I don't have much experience with numpy and it's C API please have a patience with me :-)

1. why is needed this line: `cnp.import_array()`? The numpy documentation says[1] that it needs to be used when numpy CAPI is used. But when I had a look into the generated C code function `foo()` is not using any numpy array. Instead it uses standard python API to access array members (in both 0.29.X and 3.0.X):

![image](https://github.com/cython/cython/assets/827060/7201aba4-6c11-4d36-9f47-6603824a7a5f)

2. Generated C code seems to be the same in both versions, I checked also called `__Pyx_*` functions but they seems to have same performance in both 3.0 and 0.29.X Cython.

3. When I have commented out line `cnp.import_array()` I got same performance in both 0.29.X and 3.0.X cython.

Long story short, it seems that the performance impact has the line `cnp.import_array()` but not sure why. I am even not sure whether is needed in this example.

[1] https://numpy.org/doc/1.13/reference/c-api.array.html#c.import_array

**Comment by da-woods** [comment]
> 1. why is needed this line: `cnp.import_array()`

It probably isn't needed. But we generally recommend it if there's any doubt. There's a few cases where it's needed in 3.0 and not in 0.29.x. Cython actually generates it automatically if you cimport numpy and omit it because enough people were getting mystery crashes

**Comment by da-woods** [comment]
If I had to guess (and this is a complete guess) we've made some changes to Python object indexing to support things like `__class_getitem__()`. I suspect it's this.

**Comment by matusvalo** [comment]
> If I had to guess (and this is a complete guess) we've made some changes to Python object indexing to support things like __class_getitem__().

Indeed, now I did experiment. I commented out in the loop of generated C file lines as follows:

```c
  for (__pyx_t_4 = 0; __pyx_t_4 < __pyx_t_3; __pyx_t_4+=1) {
    __pyx_v_i = __pyx_t_4;

    /* "main.pyx":10
 *     n = len(arr)
 *     for i in range(n):
 *         arr[i] += 1             # <<<<<<<<<<<<<<
 */
    __pyx_t_5 = __pyx_v_i;
    __pyx_t_6 = __Pyx_GetItemInt(((PyObject *)__pyx_v_arr), __pyx_t_5, int, 1, __Pyx_PyInt_From_int, 0, 1, 1); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 10, __pyx_L1_error)
    // __Pyx_GOTREF(__pyx_t_6);
    // __pyx_t_7 = __Pyx_PyInt_AddObjC(__pyx_t_6, __pyx_int_1, 1, 1, 0); if (unlikely(!__pyx_t_7)) __PYX_ERR(0, 10, __pyx_L1_error)
    // __Pyx_GOTREF(__pyx_t_7);
    // __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
    // if (unlikely((__Pyx_SetItemInt(((PyObject *)__pyx_v_arr), __pyx_t_5, __pyx_t_7, int, 1, __Pyx_PyInt_From_int, 0, 1, 1) < 0))) __PYX_ERR(0, 10, __pyx_L1_error)
    // __Pyx_DECREF(__pyx_t_7); __pyx_t_7 = 0;
  }
```

And it seems that 0.29.X Cython is faster then 3.0.X Cython...

**Comment by matusvalo** [comment]
> https://github.com/cython/cython/commit/c0ceabe1ee0904477ed0b0b9371ae43915d7bd2f
>
> I'd try reverting that and seeing if the performance changes (just for diagnosis... I don't think we should fix it by just reverting it)

This was introduced in cython 3.0a2 (if I am not wrong) and when I installed 3.0a1 the issue was still present... But please doublecheck me.

**Comment by da-woods** [comment]
> > [c0ceabe](https://github.com/cython/cython/commit/c0ceabe1ee0904477ed0b0b9371ae43915d7bd2f)
> > I'd try reverting that and seeing if the performance changes (just for diagnosis... I don't think we should fix it by just reverting it)
> 
> This was introduced in cython 3.0a2 (if I am not wrong) and when I installed 3.0a1 the issue was still present... But please doublecheck me.

Yeah you're right. If the issue was still in 3.0a1 then it's probably something else.

**Comment by scoder** [comment]
> If the issue was still in 3.0a1 then it's probably something else.

I think we previously looked for the sequence protocol first and only then considered mappings. That might well make a difference in the case of NumPy. This was changed to match Python's own behaviour.

**Comment by scoder** [comment]
If I comment out the entire mapping handler part in `__Pyx_GetItemInt_Fast()`, it becomes ~15% faster for me. That makes the code incorrect, but it does make a major difference in this case.
https://github.com/cython/cython/blob/9827c6085e2141db71c55ae231a4a09a878dd524/Cython/Utility/ObjectHandling.c#L445-L451

**Comment by scoder** [comment]
So, the issue here is that NumPy implements both the CPython sequence protocol (for simple indexing, length, contains, etc.) and the mapping protocol (for slicing). Since Cython 3.0, we follow Python's behaviour of looking for the mapping protocol before the sequence protocol. Previously, we passed C integer indexes directly into the sequence protocol, ignoring the mapping protocol if both were implemented.

In the specific case of NumPy, this means that we hit a double slow path. First, we have to create a new Python `int` object from the `Py_ssize_t` C index to pass it into the mapping protocol, then NumPy receives a generic object and has to determine that we're up to simple indexing, before it can unpack the index into a C integer and look up the array value.

I'm not sure how to deal with this. We're not using the buffer interface here, so the code really is slower than necessary. Should we special-case NumPy for code that could much better be accelerated differently?

**Comment by scoder** [comment]
> but there are number of places in our Cython code that need to be able to handle an arbitrary NumPy array.

Would fused types help here? That's what a couple of projects do, they just create a fused type of all possible NumPy dtypes and then let Cython generate function variants for each of them. Adds a bunch of code, and also the runtime overhead of checking the dtype and unpacking the array accordingly, but if the processing afterwards is non-trivial, or the array is somewhat large, it's probably worth it.

**Comment by scoder** [comment]
> Would fused types help here?

I just tried it on the reported example, and the calculation of a 1,000,000 item array is ~300 times faster with fused types. For 1000 items, it's still about 200 times faster.

**Comment by matusvalo** [comment]
> I'm not sure how to deal with this. We're not using the buffer interface here, so the code really is slower than necessary. Should we special-case NumPy for code that could much better be accelerated differently?

Just thinking aloud. Does it make sense to add support for abc.collections in cython?

```cython
def foo(Sequence arr):
    cdef int n, i
    n = len(arr)
    for i in range(n):
        arr[i] += 1
```

This would benefit also pure python where it will nicely:

```python
import cython
from collections.abc import Sequence

def foo(arr: Sequence):
    n: cython.int
    i: cython.int
    n = len(arr)
    for i in range(n):
        arr[i] += 1
```
[1] https://docs.python.org/3/library/collections.abc.html

**Comment by da-woods** [comment]
I don't think the distinction between the Python interfaces `Sequence` and `Mapping` is always preserved at the C API level.  E.g. for Cython extension types itself I think we always define `__getitem__` under `tp_mapping`.

I like the basic idea of being able to pick out the fast-path, but not sure how it quite works here.

**Comment by jbrockmendel** [comment]
Ages ago pandas used to have a helper `util.get_value_at` (https://github.com/pandas-dev/pandas/blob/00416816fa5be34567185149fdfcbf2d0e068af0/pandas/_libs/util.pxd#L32, and under the hood https://github.com/pandas-dev/pandas/blob/00416816fa5be34567185149fdfcbf2d0e068af0/pandas/_libs/src/numpy_helper.h#L26) that we used for `arr[i]`  At some point I removed it in favor of calling `ndarray.__getitem__` directly, but we could re-implement it.

**Comment by jbrockmendel** [comment]
> they just create a fused type of all possible NumPy dtypes

IIRC you can't put e.g. np.datetime64 dtypes into a fused type (or declare them non-fused as `ndarray[datetime64]`).  I don't think that is relevant to the OP's issue, but prevents this from being a cure-all.

**Comment by da-woods** [comment]
> Ages ago pandas used to have a helper `util.get_value_at` (https://github.com/pandas-dev/pandas/blob/00416816fa5be34567185149fdfcbf2d0e068af0/pandas/_libs/util.pxd#L32, and under the hood https://github.com/pandas-dev/pandas/blob/00416816fa5be34567185149fdfcbf2d0e068af0/pandas/_libs/src/numpy_helper.h#L26) that we used for `arr[i]` At some point I removed it in favor of calling `ndarray.__getitem__` directly, but we could re-implement it.

You could also try using the C API function [`PySequence_GetItem`](https://docs.python.org/3/c-api/sequence.html#c.PySequence_GetItem) which I think will skip directly to the fast path of the `__getitem__` lookup and be simpler than what you used to have.

**Comment by jbrockmendel** [comment]
Last thing that comes to mind: in order to handle unknown-dimension ndarrays pandas has been moving to using a pattern like

```
    it = <flatiter>PyArray_IterNew(arr)
    for i in range(n):
        val = PyArray_GETITEM(arr, PyArray_ITER_DATA(it))
```

it's not pretty, and the `__setitem__` analogue is _really_ not pretty.  Since the input arrays have unknown ndim, they can't have their types declared.  AFAIK that hasn't killed perf.  So moving to that pattern more is an option.

**Comment by scoder** [comment]
> I don't think the distinction between the Python interfaces `Sequence` and `Mapping` is always preserved at the C API level. E.g. for Cython extension types itself I think we always define `__getitem__` under `tp_mapping`.

We define both, with a slower "create PyLong, then delegate" wrapper in `sq_item`. So preferring `sq_item` would actually slow down Cython generated code.

I now wonder if we could avoid that wrapper if users defined `__getitem__` with a C integer type as index argument, so I created https://github.com/cython/cython/issues/5729 to track the idea.

**Comment by scoder** [comment]
NumPy 2.5 will apparently have the sequence type flag set: https://github.com/numpy/numpy/issues/30519
That will allow Cython to short-circuit directly to the sequence protocol if the subscript index is a C integer, as implemented in https://github.com/cython/cython/pull/7432

I think we can consider this ticket resolved.
